### PR TITLE
OOP

### PR DESCRIPTION
**Pull request by YuliiaHordiichuk**
### OOP exercise

[Demo](https://yuliiahordiichuk.github.io/a-tiny-JS-world/)
[Code base](https://github.com/YuliiaHordiichuk/a-tiny-JS-world/tree/post-oop)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/YuliiaHordiichuk/post-oop/index.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+class Animal extends Inhabitant {
+   constructor(name, gender, saying, species) {
+      super(name, gender, saying, species); 
+      this.legs = 4; 
+   }
+
+   getMessage () {
+      return `${super.getMessage()} I have ${this.legs} legs.`
+   }
```
Comment (OleksiyRudenko) [inline submissions/YuliiaHordiichuk/post-oop/index.js:52]: All inhabitants have legs. Why handling them needs a different approach compared with name, gender, species and saying?

File: submissions/YuliiaHordiichuk/post-oop/index.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+const man = new Human('Mark', 'male', 'Hi, there!', 'human'); 
+const woman = new Human('Jane', 'female', 'Hello, darling!', 'human');
```
Comment (OleksiyRudenko) [inline submissions/YuliiaHordiichuk/post-oop/index.js:56]: Every Human apparently is attributed to `human` species. Can we avoid passing it explicitly at construction time?

File: submissions/YuliiaHordiichuk/post-oop/index.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+   getMessage () {
+      return `${super.getMessage()} I have ${this.legs} legs.`
+   }
+}
+
+const man = new Human('Mark', 'male', 'Hi, there!', 'human'); 
+const woman = new Human('Jane', 'female', 'Hello, darling!', 'human'); 
+const dog = new Animal('Donny', 'male', 'Woof!', 'dog'); 
+const cat = new Animal('Beniia', 'male', 'Mrrrr!','cat')
```
Comment (OleksiyRudenko) [inline submissions/YuliiaHordiichuk/post-oop/index.js:58]: Well, humans are animals. Besides humans deserved a class for their species, while dogs and cats didn't. Why?

File: submissions/YuliiaHordiichuk/post-oop/index.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+   };
+
+   getMessage() {
+      const messageText = [
+      this.saying,
+      `I am a ${this.species}.`, 
+      `My name is ${this.name}.`,
+      `Nice to meet you!`,
+      `My gender is ${this.gender}.`
+   ];
```
Comment (OleksiyRudenko) [inline submissions/YuliiaHordiichuk/post-oop/index.js:26]: If you rename `getMessage` into `toString` you will have a magic method.
`toString` is a method that is called implictly in string context. `print` ensures string context for its arguments.

`toString` is defined for every object in JS. It is `toString` that is responsible for output of `[Object object]` in some contexts (e.g. ``console.log(`Logging an object: ${obj}`)``). It is overloadable and we can benefit from this.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@YuliiaHordiichuk nice job!
Let's make classes structure consistent (`legs` handling). This will also make code a bit shorter.
Further we can get rid of passing species names when creating inhabitants.

File: submissions/YuliiaHordiichuk/post-oop/index.js (hunk 1 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/post-oop/index.js]: You are right, done

File: submissions/YuliiaHordiichuk/post-oop/index.js (hunk 3 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/post-oop/index.js]: I`ve made different classes for cats and dogs and also added additional classes for men and women

File: submissions/YuliiaHordiichuk/post-oop/index.js (hunk 4 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/post-oop/index.js]: Didn`t know about this, thanks!

**Review by OleksiyRudenko — APPROVED** [review]
@YuliiaHordiichuk great job!

**Pull request merged**